### PR TITLE
[rpc] request for formatting dune files

### PR DESCRIPTION
**Pull request by rgrinberg**
Needed for lsp to support formatting of dune files

### CODE REVIEW

File: src/dune_rpc_impl/server.ml [hunk 1]
```diff
@@ -356,6 +358,38 @@ let handler (t : t Fdecl.t) : 'a Dune_rpc_server.Handler.t =
     in
     Handler.request rpc cb Public.Request.diagnostics
   in
+  let () =
+    let cb =
+      let f (`Path path, `Contents contents) =
+        let+ version =
+          Memo.Build.run
```
Comment (jeremiedimino) [inline src/dune_rpc_impl/server.ml:365]: I'm not sure it is safe to call `Memo.Build.run` multiple times in parallel. More precisely, errors might not go to the right place.

Comment (rgrinberg) [reply src/dune_rpc_impl/server.ml:365]: That's true. Should I use Build_system.run then?

**Comment by rgrinberg** [comment]
While I'm at it, I added a request to promote files. Will be useful for the lsp server as well.

File: src/dune_rpc_impl/server.ml (hunk 1 above)
Comment (jeremiedimino) [reply src/dune_rpc_impl/server.ml:380]: Using `Build_system.run` doesn't seem great :/ If there is a build in progress, the request is going to take forever to complete.

I re-read the code of `Memo`, and I think it should actually be ok to `Memo.run`. We should discuss it next week when @snowleopard is back, but in the meantime we can go for it.

We could also have tests for this.

**Review by jeremiedimino — APPROVED** [review]
Code looks correct to me. /cc @cwong-ocaml who might want to have a look at the new requests, for the ecaml side of things.

File: src/dune_rpc_impl/server.ml (hunk 1 above)
Comment (snowleopard) [reply src/dune_rpc_impl/server.ml:380]: If I remember correctly, concurrent calls to `Memo.run` with different error handlers may lead to subtly surprising behaviour, where only one of the handlers receives the errors "early", i.e. all but one handler may receive the errors late, after all concurrent branches have completed.

If `Memo.run` is called at "top-level", i.e. there is no outer `Memo.run_with_error_handler` that sets up a user-specified handler, then Memo uses a trivial error handler that simply ignores the errors.

See a comment next to `val run_with_error_handler` in the `mli`.

Comment (jeremiedimino) [reply src/dune_rpc_impl/server.ml:380]: Ah, we should have just read the doc!

In this case it should be fine then, as this `Memo.run` is called from the RPC which is living outside of the memo world.

Thanks @snowleopard

File: otherlibs/dune-rpc/private/dune_rpc_private.mli [hunk 2]
```diff
@@ -406,6 +406,11 @@ module Public : sig
     val ping : (unit, unit) t
 
     val diagnostics : (unit, Diagnostic.t list) t
+
+    val format_dune_file :
+      ([ `Path of string ] * [ `Contents of string ], string) t
```
Comment (cwong-ocaml) [inline otherlibs/dune-rpc/private/dune_rpc_private.mli:411]: It feels a bit inconsistent to give a regular pair of polymorphic variants here, but using regular variant types elsewhere.

Comment (rgrinberg) [reply otherlibs/dune-rpc/private/dune_rpc_private.mli:411]: I can introduce a dedicated record for this if that's what you prefer.

Comment (cwong-ocaml) [reply otherlibs/dune-rpc/private/dune_rpc_private.mli:411]: After more thought, I don't think it's a big deal. It's less cluttered to do it this way, so may as well leave it.

Comment (rgrinberg) [reply otherlibs/dune-rpc/private/dune_rpc_private.mli:411]: Actually, I like introducing a record here. I expect that it will be used in other calls.

**Comment by rgrinberg** [comment]
I added a dedicated type for paths. Should be useful as we are going to be many paths back and forth. For now I'll leave `Contents of string` as is. I don't see where else it might be handy.

**Pull request merged**